lib/client-mock.js

```diff
@@ -0,0 +1,303 @@
+'use strict'
+
+const Client = require('./core/client')
+const { kUrl } = require('./core/symbols')
+
+const originalDispatch = Client.prototype.dispatch
+
+const mockDispatches = []
```

Comment:
Are we sure we want these to be global and not isntance-specific? I think it might be more flexible long term.

Reply:
Yeah I agree! I was initially concerned about the concurrency impact of overriding Client.prototype.dispatch which I why I chose this pattern in the first place. I've done some experiments just now after your suggestion with jest and tap and confirmed that this should not be a problem as tests within the same file should be run in order and those in different files run in parallel (in different processes). I will refactor to make this instance specific on each `MockClient` :)
<img width="870" alt="Screenshot 2021-03-11 at 22 54 19" src="https://user-images.githubusercontent.com/31290910/110866230-b9f16680-82bc-11eb-9081-fcfcc1916b8b.png">

Reply:
Functionally, I think we're pretty much there now. I'm going to tidy up the work I've done so far, add your suggested refactors, then typings and documentation